src/Listener/Form/Listener.php

```diff
@@ -27,7 +34,60 @@ class Listener extends Extension
      * @param $vars
      * @param Form $form
      */
-    public function afterCallFormHandler(HTTPRequest $request, $funcName, $vars, $form): void
+    public function afterCallFormHandlerController(HTTPRequest $request, $funcName, $vars, $form): void
+    {
+        $this->triggerAction($request, $funcName, $vars, $form);
+    }
+
+    /**
+     * Extension point in @see FormRequestHandler::httpSubmission
+     * form handler action via form submission action
+     *
+     * @param HTTPRequest $request
+     * @param $funcName
+     * @param $vars
+     * @param Form $form
+     */
+    public function afterCallFormHandlerMethod(HTTPRequest $request, $funcName, $vars, $form): void
+    {
+        $this->triggerAction($request, $funcName, $vars, $form);
+    }
+
+    /**
+     * Extension point in @see FormRequestHandler::httpSubmission
+     * form method action via form submission action
+     *
+     * @param HTTPRequest $request
+     * @param $funcName
+     * @param $vars
+     * @param Form $form
+     */
+    public function afterCallFormHandlerFormMethod(HTTPRequest $request, $funcName, $vars, $form): void
+    {
+        $this->triggerAction($request, $funcName, $vars, $form);
+    }
+
+    /**
+     * Extension point in @see FormRequestHandler::httpSubmission
+     * form field method action via form submission action
+     *
+     * @param HTTPRequest $request
+     * @param $funcName
+     * @param $vars
+     * @param Form $form
+     */
+    public function afterCallFormHandlerFieldMethod(HTTPRequest $request, $funcName, $vars, $form): void
+    {
+        $this->triggerAction($request, $funcName, $vars, $form);
+    }
+
+    /**
+     * @param HTTPRequest $request
+     * @param $funcName
+     * @param $vars
+     * @param $form
+     */
+    private function triggerAction(HTTPRequest $request, $funcName, $vars, $form): void
```

Comment:
Correcting form submission extension points. All four must be covered and can't be easily replaced with a generic extension point. See comments in the doc block for more information.

Reply:
I still don't understand why we need four extension points. See my comment here: https://github.com/silverstripe/silverstripe-framework/pull/9340#pullrequestreview-339525999

I understand there are different conditions created by how the form is handled, but they all can provide a `funcName` for the action, and so long as we provide a the `$subject` to the extension hook, whether that is `Controller`, `Form`, `Field`, etc., the developer should have plenty to work with.

Reply:
Oh, I see. I missed your commits in the other repo. Will update my PR :)

Reply:
Change reverted.